liquidctl/driver/asetek_pro.py

```diff
@@ -0,0 +1,256 @@
+_PRO_CMD_READ_PUMP_SPEED = 0x31
+_PRO_CMD_READ_PUMP_MODE = 0x33
+_PRO_CMD_WRITE_PUMP_MODE = 0x32
+_PRO_CMD_WRITE_COLOR_LIST = 0x56
+_PRO_PUMP_MODES = [
+    'Quiet',
+    'Balanced',
+    'Performance'
+    ]
+
+_PRO_SPEEDS = {
+    'rainbow': [
+        0x30, # slow
+        0x18, # medium
+        0x0C  # fast
+    ],
+    'shift': [
+        0x46, # slow
+        0x28, # medium
+        0x0F  # fast
+    ],
+    'pulse':[
+        0x50, # slow
+        0x37, # medium
+        0x1E  # fast
+    ],
+    'blinking': [
+        0x0F, # slow
+        0x0A, # medium
+        0x05  # fast
+    ]
+}
+
+_PRO_COLOR_CHANGE_MODES = {
+    'rainbow': [0x55, 0x01],
+    'alert': [],
+    'shitf': [0x55, 0x01],
+    'pulse': [0x52, 0x01],
+    'blinking': [0x58, 0x01],
+    'fixed': [0x55, 0x01]
+}
+
+
+class CorsairAsetekProDriver(Modern690Lc):
+    """liquidctl driver for Corsair-branded sixth generation Asetek coolers."""
+    SUPPORTED_DEVICES = [
+        (0x1b1c, 0x0c12, None, 'Corsair Hydro H150i Pro', {}),
+        (0x1b1c, 0x0c13, None, 'Corsair Hydro H115i Pro', {}),
+        (0x1b1c, 0x0c14, None, 'Corsair Hydro H110i Pro', {}),
+        (0x1b1c, 0x0c15, None, 'Corsair Hydro H100i Pro', {}),
+        (0x1b1c, 0x0c16, None, 'Corsair Hydro H80i Pro', {}),
+    ]
+
+    def _write(self, data):
+        """Write data to the AIO and log"""
+        _LOGGER.debug('write %s', LazyHexRepr(data))
+        self.device.write(_PRO_WRITE_ENDPOINT, data, _WRITE_TIMEOUT)
+
+    # Not calling this (or begin transaction) since they do not seem to be needed for H100i Pro
+    def _end_transaction_and_read(self, length=_READ_LENGTH):
+        """End the transaction by reading from the device.
+        According to the official documentation, as well as Craig's open-source
+        implementation (libSiUSBXp), it should be necessary to check the queue
+        size and read data in chunks.  However, leviathan and its derivatives
+        seem to work fine without this complexity; we currently try the same
+        approach.
+        """
+        msg = self.device.read(_PRO_READ_ENDPOINT, length, _READ_TIMEOUT)
+        _LOGGER.debug('received %s', LazyHexRepr(msg))
+        self.device.release()
+        return msg
+
+    def _write_color_change_speed(self, mode, speed):
+        """Send the speed to cycle colors on the RGB pump"""
+        if speed < 1 or speed > 3:
+            raise ValueError('Unsupported speed ({}), only 1-3 supported'.format(speed))
+
+        self._write([0x53, _PRO_SPEEDS[mode][int(speed)-1]])
+        self._end_transaction_and_read(3)
+
+    def _write_color(self, colors):
+        """Write a list of colors to cycle between, the cycle pattern comes from the
+        value used for cycle speed"""
+        colorCount = 0
+        setColors = []
+        for color in colors:
+            colorCount = colorCount + 1
+            setColors = setColors + color
+        self._write([_PRO_CMD_WRITE_COLOR_LIST, colorCount] + setColors)
```

Comment:
Assuming `colors` is indeed a list, this entire chunk should be equivalent to:

```
self._write([_PRO_CMD_WRITE_COLOR_LIST, len(colors)] + colors)
```

That said, this would try to write a list of lists, which is probably not the intended behavior (and could explain why this part of the driver isn't working).  Instead it may be necessary to use `itertools.chain`.

Reply:
I've moved this to use `itertools.chain` as it's neater, but it's functional behavior hasn't changed.